### PR TITLE
Backport SHA-256 assembly optimisations

### PR DESCRIPTION
**Pull request by str4d**
Cherry-picked from the following upstream PRs:
- bitcoin/bitcoin#10821
- bitcoin/bitcoin#10971
- bitcoin/bitcoin#11176
- bitcoin/bitcoin#13191
- bitcoin/bitcoin#13393
- bitcoin/bitcoin#13438
- bitcoin/bitcoin#13471
- bitcoin/bitcoin#13386
- bitcoin/bitcoin#13611
- bitcoin/bitcoin#13788
- bitcoin/bitcoin#18553

Closes zcash/zcash#5408.

### CODE REVIEW

**Comment by str4d** [comment]
With the SHA-256 assembly optimisations, the big-block benchmarks decreased by (on my Ryzen 9 5950X):
- Transparent:
  - Uncached: 37.4%
  - Cached: 25.9%
- Sapling:
  - Uncached: 4.2%
  - Cached: 2.8%
- Orchard:
  - Uncached: 10%
  - Cached: 5.3%

<details>
<summary>Data</summary>

In the measurements below, the first result is uncached, and the remaining 9 are cached (due to a happy bug in how `src/zcbenchmarks.cpp` uses `ConnectBlock` with `fJustCheck = true`, causing the cache to be populated for transparent foverver and for Sapling/Orchard since #6073 merged).

### 420f8dfe38fd6b2465a665324366c2ae14aa98f4

```
❯ ./qa/zcash/performance-measurements.sh time connectblockslow
block-107134.tar.xz: OK
[
  {
    "runningtime": 0.150669
  },
  {
    "runningtime": 0.09304999999999999
  },
  {
    "runningtime": 0.09410300000000001
  },
  {
    "runningtime": 0.09286899999999999
  },
  {
    "runningtime": 0.094154
  },
  {
    "runningtime": 0.09192
  },
  {
    "runningtime": 0.092637
  },
  {
    "runningtime": 0.09251
  },
  {
    "runningtime": 0.093087
  },
  {
    "runningtime": 0.09178500000000001
  }
]

❯ ./qa/zcash/performance-measurements.sh time connectblocksapling
block-1723244.tar.xz: OK
[
  {
    "runningtime": 1.923672
  },
  {
    "runningtime": 0.450483
  },
  {
    "runningtime": 0.446739
  },
  {
    "runningtime": 0.4474050000000001
  },
  {
    "runningtime": 0.445815
  },
  {
    "runningtime": 0.445268
  },
  {
    "runningtime": 0.44584
  },
  {
    "runningtime": 0.445408
  },
  {
    "runningtime": 0.446498
  },
  {
    "runningtime": 0.445251
  }
]

❯ ./qa/zcash/performance-measurements.sh time connectblockorchard
block-1708048.tar.xz: OK
[
  {
    "runningtime": 0.475553
  },
  {
    "runningtime": 0.175597
  },
  {
    "runningtime": 0.175143
  },
  {
    "runningtime": 0.181652
  },
  {
    "runningtime": 0.181104
  },
  {
    "runningtime": 0.181345
  },
  {
    "runningtime": 0.180676
  },
  {
    "runningtime": 0.180846
  },
  {
    "runningtime": 0.182954
  },
  {
    "runningtime": 0.181823
  }
]
```

### This branch

```
❯ ./qa/zcash/performance-measurements.sh time connectblockslow
block-107134.tar.xz: OK
[
  {
    "runningtime": 0.09435499999999999
  },
  {
    "runningtime": 0.067063
  },
  {
    "runningtime": 0.069358
  },
  {
    "runningtime": 0.069522
  },
  {
    "runningtime": 0.068615
  },
  {
    "runningtime": 0.071535
  },
  {
    "runningtime": 0.06694600000000001
  },
  {
    "runningtime": 0.070537
  },
  {
    "runningtime": 0.06866700000000001
  },
  {
    "runningtime": 0.06765599999999999
  }
]

❯ ./qa/zcash/performance-measurements.sh time connectblocksapling
block-1723244.tar.xz: OK
[
  {
    "runningtime": 1.843351
  },
  {
    "runningtime": 0.434795
  },
  {
    "runningtime": 0.434568
  },
  {
    "runningtime": 0.433503
  },
  {
    "runningtime": 0.434782
  },
  {
    "runningtime": 0.434338
  },
  {
    "runningtime": 0.433024
  },
  {
    "runningtime": 0.4331930000000001
  },
  {
    "runningtime": 0.434772
  },
  {
    "runningtime": 0.434312
  }
]

❯ ./qa/zcash/performance-measurements.sh time connectblockorchard
block-1708048.tar.xz: OK
[
  {
    "runningtime": 0.428119
  },
  {
    "runningtime": 0.165134
  },
  {
    "runningtime": 0.164023
  },
  {
    "runningtime": 0.163555
  },
  {
    "runningtime": 0.170682
  },
  {
    "runningtime": 0.17297
  },
  {
    "runningtime": 0.172324
  },
  {
    "runningtime": 0.170939
  },
  {
    "runningtime": 0.177271
  },
  {
    "runningtime": 0.178633
  }
]
```

</details>

**Comment by str4d** [comment]
The micro-benchmarks for this branch:
```
#Benchmark,     count, min(ns), max(ns), average(ns), min_cycles, max_cycles, average_cycles
SHA256,          2560,  426209,  430203,      426918,    1446387,    1459940,        1448792
SHA256D64_1024, 24576,   42870,   42969,       42903,     145485,     145822,         145597
```

**Comment by str4d** [comment]
Rebased on master now that #6079 has merged.

**Comment by str4d** [comment]
I've now been through all of the `src/crypto/sha256*` files, and this PR contains every meaningful change and bugfix to the files introduced by this PR. The remaining SHA-256 changes from upstream (e.g. ARM intrinsics) are tracked in #6080, which blocks on #5407 due to merge conflicts).

**Review by nuttycom — DISMISSED** [review]
utACK. I verified each commit and checked that the Zcash-specific changes made sense, but did not closely inspect the logic from upstream.

File: src/crypto/sha256_sse4.cpp [hunk 1]
```diff
@@ -0,0 +1,1506 @@
+// Copyright (c) 2017 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
```
Comment (daira) [inline src/crypto/sha256_sse4.cpp:3]:
```suggestion
// file COPYING or https://www.opensource.org/licenses/mit-license.php .
```

File: src/crypto/sha256_shani.cpp [hunk 2]
```diff
@@ -0,0 +1,356 @@
+// Copyright (c) 2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
```
Comment (daira) [inline src/crypto/sha256_shani.cpp:3]:
```suggestion
// file COPYING or https://www.opensource.org/licenses/mit-license.php .
```

File: src/crypto/sha256_sse4.cpp [hunk 3]
```diff
@@ -0,0 +1,1506 @@
...
+; Linux:    yasm -f x64 -f elf64 -X gnu -g dwarf2 -D LINUX -o sha256_sse4.o sha256_sse4.asm
+;
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;
+; This code is described in an Intel White-Paper:
+; "Fast SHA-256 Implementations on Intel Architecture Processors"
+;
+; To find it, surf to http://www.intel.com/p/en_US/embedded 
+; and search for that title.
+; The paper is expected to be released roughly at the end of April, 2012
```
Comment (daira) [inline src/crypto/sha256_sse4.cpp:1006]:
```suggestion
; at <https://www.intel.com/content/dam/www/public/us/en/documents/white-papers/sha-256-implementations-paper.pdf>.
```
(The described search does not find it.)

**Review by daira — COMMENTED** [review]
Flushing comments; not finished reviewing yet.

File: src/bench/merkle_root.cpp [hunk 4]
```diff
@@ -0,0 +1,26 @@
+// Copyright (c) 2016 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
```
Comment (daira) [inline src/bench/merkle_root.cpp:3]:
```suggestion
// file COPYING or https://www.opensource.org/licenses/mit-license.php .
```

**Review by daira — APPROVED** [review]
utACK with minor suggestions.

**Review by daira — APPROVED** [review]
utACK 26d41818ad8f252f9e97c3e1553890ff5b86d34d.

I have not checked this code against upstream; I'm relying on @str4d and @nuttycom having done that.

**Review by daira — APPROVED** [review]
Checked that the SHA-256 code itself matches upstream (master), modulo changes we are intentionally not pulling in such as the ARM asm support.

**Pull request merged**